File: src/nl/rubensten/texifyidea/editor/TeXiFyProjectViewNodeDecorator.java
package nl.rubensten.texifyidea.editor;

import com.intellij.ide.projectView.PresentationData;
import com.intellij.ide.projectView.ProjectViewNode;
import com.intellij.ide.projectView.ProjectViewNodeDecorator;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.packageDependencies.ui.PackageDependenciesNode;
import com.intellij.ui.ColoredTreeCellRenderer;
import nl.rubensten.texifyidea.TexifyIcons;

import javax.swing.*;
import java.util.HashMap;
import java.util.Map;

/**
 * @author Ruben Schellekens
 */
public class TeXiFyProjectViewNodeDecorator implements ProjectViewNodeDecorator {

    private static final Map<String, Icon> FILE_ICONS = new HashMap<>();
    static {
        FILE_ICONS.put("pdf", TexifyIcons.PDF_FILE);
        FILE_ICONS.put("dvi", TexifyIcons.DVI_FILE);
    }

    private void setIcon(ProjectViewNode projectViewNode, PresentationData presentationData) {
        VirtualFile file = projectViewNode.getVirtualFile();
        if (file == null) {
            return;
        }
        if (file.isDirectory()) {
            return;
        }

        String extension = file.getExtension();
        if (extension == null) {
            return;
        }

        Icon icon = FILE_ICONS.get(file.getExtension().toLowerCase());
        if (icon == null) {
            return;
        }

        presentationData.setIcon(icon);
    }

    @Override
    public void decorate(ProjectViewNode projectViewNode, PresentationData presentationData) {
        setIcon(projectViewNode, presentationData);
    }

    @Override
    public void decorate(PackageDependenciesNode packageDependenciesNode, ColoredTreeCellRenderer coloredTreeCellRenderer) {
        // Do nothing.
    }
}

File: src/nl/rubensten/texifyidea/project/TeXiFyProjectViewNodeDecorator.kt
package nl.rubensten.texifyidea.project

import com.intellij.ide.projectView.PresentationData
import com.intellij.ide.projectView.ProjectViewNode
import com.intellij.ide.projectView.ProjectViewNodeDecorator
import com.intellij.packageDependencies.ui.PackageDependenciesNode
import com.intellij.ui.ColoredTreeCellRenderer
import nl.rubensten.texifyidea.util.Magic

/**
 * @author Ruben Schellekens
 */
class TeXiFyProjectViewNodeDecorator : ProjectViewNodeDecorator {

    private fun setIcon(projectViewNode: ProjectViewNode<*>, presentationData: PresentationData) {
        val file = projectViewNode.virtualFile ?: return
        if (file.isDirectory) {
            return
        }

        var extension = file.extension ?: return
        if (extension == "gz") {
            extension = "synctex.gz"

            if (!file.name.toLowerCase().endsWith("synctex.gz")) {
                return
            }
        }

        val icon = Magic.Icon.fileIcons[extension.toLowerCase()] ?: return

        presentationData.setIcon(icon)
    }

    override fun decorate(projectViewNode: ProjectViewNode<*>, presentationData: PresentationData) {
        setIcon(projectViewNode, presentationData)
    }

    override fun decorate(packageDependenciesNode: PackageDependenciesNode, coloredTreeCellRenderer: ColoredTreeCellRenderer) {
        // Do nothing.
    }
}
